Mage/src/main/java/mage/abilities/common/CaseAbility.java

```diff
@@ -0,0 +1,213 @@
+
+        if (solvedAbility instanceof TriggeredAbility) {
+            addSubAbility(new CaseTriggeredAbility((TriggeredAbility) solvedAbility, true));
+        } else if (solvedAbility instanceof ActivatedAbility) {
+            addSubAbility(new CaseActivatedAbility((ActivatedAbility) solvedAbility, true));
+        } else if (solvedAbility instanceof StaticAbility) {
+            addSubAbility(new CaseStaticAbility((StaticAbility) solvedAbility, true));
+        } else {
+            throw new IllegalArgumentException("Solved Case ability supports only triggered, activated and static abilities");
+        }
+    }
+
+    protected CaseAbility(final CaseAbility ability) {
+        super(ability);
+    }
+
+    @Override
+    public CaseAbility copy() {
+        return new CaseAbility(this);
+    }
+
+    private static String removeIf(String text) {
+        if (text.startsWith("if ")) {
+            return text.substring(3);
+        }
+        return text;
+    }
+}
+
+enum TrueCondition implements Condition {
+    instance;
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        return true;
+    }
+}
+
+class CaseTriggeredAbility extends ConditionalTriggeredAbility {
+
+    CaseTriggeredAbility(TriggeredAbility ability) {
+        this(ability, false);
+    }
+
+    CaseTriggeredAbility(TriggeredAbility ability, boolean solved) {
+        super(ability, solved ? SolvedSourceCondition.instance : TrueCondition.instance,
+                (solved ? "Solved &mdash; " : "") + CardUtil.getTextWithFirstCharUpperCase(ability.getRule("this Case")));
+    }
+
+    private CaseTriggeredAbility(final CaseTriggeredAbility ability) {
+        super(ability);
+    }
+
+    @Override
+    public CaseTriggeredAbility copy() {
+        return new CaseTriggeredAbility(this);
+    }
+}
+
+class CaseActivatedAbility extends ConditionalActivatedAbility {
+
+    CaseActivatedAbility(ActivatedAbility ability) {
+        this(ability, false);
+    }
+
+    CaseActivatedAbility(ActivatedAbility ability, boolean solved) {
+        super(ability.getZone(), ability.getEffects().get(0), ability.getCosts(),
+                solved ? SolvedSourceCondition.instance : TrueCondition.instance,
+                (solved ? "Solved &mdash; " : "") + CardUtil.getTextWithFirstCharUpperCase(ability.getRule("this Case")));
+    }
+
+    private CaseActivatedAbility(final CaseActivatedAbility ability) {
+        super(ability);
+    }
+
+    @Override
+    public CaseActivatedAbility copy() {
+        return new CaseActivatedAbility(this);
+    }
+}
+
+class CaseStaticAbility extends StaticAbility {
+
+    CaseStaticAbility(StaticAbility ability) {
+        this(ability, false);
+    }
+
+    CaseStaticAbility(StaticAbility ability, boolean solved) {
+        super(ability.getZone(), null);
+
+        boolean first = true;
+        for (Effect effect : ability.getEffects()) {
```

Comment:
Looks very buggy in theory. You cant move effects from one ability to another — some effect logic can depends on the source (ability) and other inner data (example: dynamic targets). So it will be missed after move/extract. Also ability’s watchers can be missed (maybe).

Developer must control ability/effect/conditional itself. So I recommends to use a standard code style like “add normal ability with conditional to a card” instead container style like “take ability, extract effects and inject conditional to it”.

If this logic was required by some technical limitations, then it is necessary to justify the decision.

Reply:
I agree with this recommendation.

I understand the appeal of having `CaseAbility` handle it internally, but given the variety of places where conditionals might need to apply and the pitfalls of copying data from one effect to another, it might end up being a better approach to just require that the `SolvedSourceCondition` be added to the ability/effect before passing it into the CaseAbility.